### ISSUE TITLE
improve strlcpy handling for newer glibc & musl

### ISSUE THREAD

**Issue by robbat2** [radvd-project/radvd#248]
Distro bug https://bugs.gentoo.org/show_bug.cgi?id=947276

Make configure check libc (glibc >= 2.38, musl) for strlcpy, with fallback to libbsd.

**Comment by angus19** [comment]
Meet same fatal error during cross compiling and there could be no bsd library in such scenario.

> In file included from log.c:16:
includes.h:30:10: fatal error: bsd/string.h: No such file or directory
   30 | #include <bsd/string.h> // strlcpy
      |          ^~~~~~~~~~~~~~
compilation terminated.

**Comment by robbat2** [comment]
> Meet same fatal error during cross compiling and there could be no bsd library in such scenario.
You can cross-compile libbsd quite fine; but that's a good reason as well to improve here and support strlcpy from new glibc & musl.

Patches welcome - likely should test libc for strlcpy, and then fall back to libbsd

**Comment by angus19** [comment]
What about including strlcpy.c - the source file and compiling/using it when strlcpy cannot be found from libc or no libbsd is installed? This is an idea. My knowledge to patch files associated with ./configure is limited for now.

**Comment by robbat2** [comment]
I think that is likely a good further improvement on top of my existing PR; hoping to have that merged for now.
https://github.com/openbsd/src/blob/master/lib/libc/string/strlcpy.c

**Comment by jlm** [comment]
I should have read this issue before submitting my pull request.

**Comment by stappersg** [comment]
On Wed, Jan 15, 2025 at 11:55:57AM -0800, John Messenger wrote:
> I should have read this issue before submitting my pull request.

No worries, I have closed #258.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by robbat2** [comment]
@stappersg  still waiting for an approval on https://github.com/radvd-project/radvd/pull/256 - I want to merge that, and add a third fallback of FreeBSD's `strlcpy.c` afterwards

**Comment by stappersg** [comment]
On Thu, Jan 16, 2025 at 04:18:09PM -0800, Robin H. Johnson wrote:
> @stappersg  still waiting for an approval on
> https://github.com/radvd-project/radvd/pull/256 - I want to merge that,
> and add a third fallback of FreeBSD's `strlcpy.c` afterwards
> 


We can do better.


Groeten
Geert Stappers
-- 
Silence is hard to parse
